src/mlpack/methods/ann/layer/convolution_impl.hpp

```diff
@@ -314,62 +306,90 @@ void Convolution<
+                           inputHeight, inSize * batchSize, false, false);
+  }
 
   gradient.set_size(weights.n_elem, 1);
   gradientTemp = arma::Cube<eT>(gradient.memptr(), weight.n_rows,
       weight.n_cols, weight.n_slices, false, false);
   gradientTemp.zeros();
-
-  for (size_t outMap = 0, outMapIdx = 0, batchCount = 0; outMap <
-      outSize * batchSize; outMap++)
+  #pragma omp parallel for
+  for (omp_size_t outMapIdx = 0; outMapIdx < outSize; outMapIdx++)
   {
-    if (outMap != 0 && outMap % outSize == 0)
+    for (size_t inMap = 0; inMap < inSize; inMap++)
     {
-      batchCount++;
-      outMapIdx = 0;
+      arma::Mat<eT> &curGradTemp = gradientTemp.slice(outMapIdx+inMap);
+      for (size_t batchCount = 0; batchCount < batchSize; batchCount++) {
+        size_t outMap = outMapIdx + batchCount*outSize;
+        arma::Mat<eT> &inputSlice = inputTemp.slice(inMap+(batchCount*inSize));
+        arma::Mat<eT> &deltaSlice = mappedError.slice(outMap);
+        arma::Mat<eT> output;
+        GradientConvolutionRule::Convolution(inputSlice, deltaSlice,
+                                            output, strideWidth, strideHeight);
+        if (gradientTemp.n_rows < output.n_rows ||
+           gradientTemp.n_cols < output.n_cols)
+        {
+         curGradTemp += output.submat(0, 0, gradientTemp.n_rows-1,
+                                      gradientTemp.n_cols-1);
+        }
+        else if (gradientTemp.n_rows > output.n_rows ||
+                gradientTemp.n_cols > output.n_cols)
+        {
+         curGradTemp.submat(0, 0, output.n_rows-1, output.n_cols-1) += output;
+        }
+        else
+        {
+         curGradTemp += output;
+        }
+        if (inMap == inSize-1) {
+          gradient.submat(weight.n_elem+(outMap%outSize), 0,
+                          weight.n_elem+(outMap%outSize),
+                          0) = arma::accu(mappedError.slice(outMap));
+        }
+      }
     }
-
+  }
+  /*for (size_t outMap = 0; outMap < outSize * batchSize; outMap++)
+  {
+    size_t outMapIdx = (outMap % outSize) * inSize,
+           batchCount = outMap/outSize;
     for (size_t inMap = 0; inMap < inSize; inMap++, outMapIdx++)
     {
-      arma::Mat<eT> inputSlice;
-      if (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0)
-      {
-        inputSlice = inputPaddedTemp.slice(inMap + batchCount * inSize);
-      }
-      else
-      {
-        inputSlice = inputTemp.slice(inMap + batchCount * inSize);
-      }
-
-      arma::Mat<eT> deltaSlice = mappedError.slice(outMap);
-
+      arma::Mat<eT> &inputSlice = inputTemp.slice(inMap+(batchCount*inSize));
+      arma::Mat<eT> &deltaSlice = mappedError.slice(outMap);
       arma::Mat<eT> output;
       GradientConvolutionRule::Convolution(inputSlice, deltaSlice,
-          output, strideWidth, strideHeight);
+                                           output, strideWidth, strideHeight);
```

Comment:
Actually, according to the [style guide](https://github.com/mlpack/mlpack/wiki/DesignGuidelines) the indentation was correct:)